File: packages/demo-source1/src/game-events.ts
import { BitReader } from "./bit-reader.js";

export const gameEventFieldTypes = {
  string: 1,
  float: 2,
  long: 3,
  short: 4,
  byte: 5,
  boolean: 6,
  uint64: 7,
} as const;

export type GameEventFieldType = keyof typeof gameEventFieldTypes;
export type GameEventValue = boolean | number | string;

export interface GameEventFieldSchema {
  readonly name: string;
  readonly type: GameEventFieldType;
}

export interface GameEventSchema {
  readonly id: number;
  readonly name: string;
  readonly fields: readonly GameEventFieldSchema[];
}

export interface DecodedGameEvent {
  readonly id: number;
  readonly name: string;
  readonly fields: Readonly<Record<string, GameEventValue>>;
  readonly schema: GameEventSchema;
}

export interface GameEventDecodeLimits {
  readonly maxEvents?: number;
  readonly maxFieldsPerEvent?: number;
  readonly maxStringBytes?: number;
}

export interface EventFieldAvailability<T extends GameEventValue> {
  readonly availability: "observed" | "unavailable";
  readonly value?: T;
  readonly provenance?: {
    readonly message: "svc_GameEvent";
    readonly eventId: number;
    readonly field: string;
  };
  readonly reason?: string;
}

export type RequiredGameEventName =
  | "weapon_fire"
  | "player_hurt"
  | "player_death";

export interface RequiredGameEventProjection {
  readonly name: RequiredGameEventName;
  readonly eventId: number;
  readonly tick: number;
  readonly actorUserId: EventFieldAvailability<number>;
  readonly victimUserId: EventFieldAvailability<number>;
  readonly attackerUserId: EventFieldAvailability<number>;
  readonly weapon: EventFieldAvailability<string>;
  readonly damage: EventFieldAvailability<number>;
  readonly health: EventFieldAvailability<number>;
  readonly decoded: DecodedGameEvent;
}

/** Decodes the bounded payload of svc_GameEventList for L4D2 protocol 2100. */
export function decodeL4d2GameEventList(
  bytes: Uint8Array,
  bitLength: number,
  eventCount: number,
  limits: GameEventDecodeLimits = {},
): readonly GameEventSchema[] {
  const maxEvents = positiveLimit(limits.maxEvents, 512, "maxEvents");
  const maxFields = positiveLimit(
    limits.maxFieldsPerEvent,
    128,
    "maxFieldsPerEvent",
  );
  const maxStringBytes = positiveLimit(
    limits.maxStringBytes,
    4_096,
    "maxStringBytes",
  );
  requireBitLength(bytes, bitLength);
  if (
    !Number.isSafeInteger(eventCount) ||
    eventCount < 0 ||
    eventCount > maxEvents
  )
    throw new RangeError(`game event count ${eventCount} is invalid`);
  const reader = new BitReader(bytes);
  const schemas: GameEventSchema[] = [];
  const ids = new Set<number>();
  for (let index = 0; index < eventCount; index += 1) {
    requireWithin(reader, bitLength, 9);
    const id = reader.readBits(9);
    if (ids.has(id)) throw new RangeError(`duplicate game event id ${id}`);
    ids.add(id);
    const name = readBoundedString(reader, bitLength, maxStringBytes);
    const fields: GameEventFieldSchema[] = [];
    while (true) {
      requireWithin(reader, bitLength, 3);
      const typeId = reader.readBits(3);
      if (typeId === 0) break;
      if (fields.length >= maxFields)
        throw new RangeError(`game event ${id} field limit exceeded`);
      const type = fieldType(typeId);
      const fieldName = readBoundedString(reader, bitLength, maxStringBytes);
      if (fieldName === "")
        throw new RangeError(`game event ${id} has an empty field name`);
      if (fields.some(({ name: value }) => value === fieldName))
        throw new RangeError(`game event ${id} repeats field ${fieldName}`);
      fields.push({ name: fieldName, type });
    }
    schemas.push({ id, name, fields });
  }
  if (reader.bitOffset !== bitLength)
    throw new RangeError(
      `game event list consumed ${reader.bitOffset} of ${bitLength} bits`,
    );
  return schemas;
}

/** Decodes one bounded svc_GameEvent payload using its preceding list schema. */
export function decodeL4d2GameEvent(
  bytes: Uint8Array,
  bitLength: number,
  schemas: ReadonlyMap<number, GameEventSchema>,
  limits: GameEventDecodeLimits = {},
): DecodedGameEvent {
  const maxStringBytes = positiveLimit(
    limits.maxStringBytes,
    4_096,
    "maxStringBytes",
  );
  requireBitLength(bytes, bitLength);
  const reader = new BitReader(bytes);
  requireWithin(reader, bitLength, 9);
  const id = reader.readBits(9);
  const schema = schemas.get(id);
  if (!schema)
    throw new RangeError(`game event ${id} has no registered schema`);
  const fields: Record<string, GameEventValue> = {};
  for (const field of schema.fields) {
    fields[field.name] = readValue(
      reader,
      bitLength,
      field.type,
      maxStringBytes,
    );
  }
  if (reader.bitOffset !== bitLength)
    throw new RangeError(
      `game event ${id} consumed ${reader.bitOffset} of ${bitLength} bits`,
    );
  return { id, name: schema.name, fields, schema };
}

/** Retains only the latest bounded schema registry while streaming a demo. */
export class L4d2GameEventDecoder {
  readonly #limits: GameEventDecodeLimits;
  #schemas = new Map<number, GameEventSchema>();

  constructor(limits: GameEventDecodeLimits = {}) {
    this.#limits = limits;
  }

  get schemas(): ReadonlyMap<number, GameEventSchema> {
    return this.#schemas;
  }

  registerList(bytes: Uint8Array, bitLength: number, eventCount: number): void {
    const decoded = decodeL4d2GameEventList(
      bytes,
      bitLength,
      eventCount,
      this.#limits,
    );
    this.#schemas = new Map(decoded.map((schema) => [schema.id, schema]));
  }

  decode(bytes: Uint8Array, bitLength: number): DecodedGameEvent {
    return decodeL4d2GameEvent(bytes, bitLength, this.#schemas, this.#limits);
  }
}

/** Projects only review-relevant events; absent protocol fields stay explicit. */
export function projectRequiredGameEvent(
  event: DecodedGameEvent,
  tick: number,
): RequiredGameEventProjection | null {
  if (!isRequiredName(event.name)) return null;
  return {
    name: event.name,
    eventId: event.id,
    tick,
    actorUserId: numberField(event, "userid"),
    victimUserId:
      event.name === "player_hurt" || event.name === "player_death"
        ? numberField(event, "userid")
        : unavailable("event has no victim role"),
    attackerUserId:
      event.name === "player_hurt" || event.name === "player_death"
        ? numberField(event, "attacker")
        : unavailable("event has no attacker role"),
    weapon: stringField(event, "weapon"),
    damage:
      event.name === "player_hurt"
        ? numberField(event, "dmg_health")
        : unavailable("event has no damage field"),
    health:
      event.name === "player_hurt"
        ? numberField(event, "health")
        : unavailable("event has no health field"),
    decoded: event,
  };
}

function numberField(
  event: DecodedGameEvent,
  field: string,
): EventFieldAvailability<number> {
  const value = event.fields[field];
  return typeof value === "number"
    ? observed(event, field, value)
    : unavailable(`schema does not expose numeric ${field}`);
}

function stringField(
  event: DecodedGameEvent,
  field: string,
): EventFieldAvailability<string> {
  const value = event.fields[field];
  return typeof value === "string"
    ? observed(event, field, value)
    : unavailable(`schema does not expose string ${field}`);
}

function observed<T extends GameEventValue>(
  event: DecodedGameEvent,
  field: string,
  value: T,
): EventFieldAvailability<T> {
  return {
    availability: "observed",
    value,
    provenance: { message: "svc_GameEvent", eventId: event.id, field },
  };
}

function unavailable<T extends GameEventValue>(
  reason: string,
): EventFieldAvailability<T> {
  return { availability: "unavailable", reason };
}

function readValue(
  reader: BitReader,
  bitLength: number,
  type: GameEventFieldType,
  maxStringBytes: number,
): GameEventValue {
  switch (type) {
    case "string":
      return readBoundedString(reader, bitLength, maxStringBytes);
    case "float":
      requireWithin(reader, bitLength, 32);
      return reader.readFloat32();
    case "long":
      requireWithin(reader, bitLength, 32);
      return reader.readSignedBits(32);
    case "short":
      requireWithin(reader, bitLength, 16);
      return reader.readSignedBits(16);
    case "byte":
      requireWithin(reader, bitLength, 8);
      return reader.readBits(8);
    case "boolean":
      requireWithin(reader, bitLength, 1);
      return reader.readBoolean();
    case "uint64": {
      requireWithin(reader, bitLength, 64);
      const low = BigInt(reader.readBits(32));
      const high = BigInt(reader.readBits(32));
      return ((high << 32n) | low).toString(10);
    }
  }
}

function readBoundedString(
  reader: BitReader,
  bitLength: number,
  maxBytes: number,
): string {
  const availableBytes = Math.floor((bitLength - reader.bitOffset) / 8);
  if (availableBytes <= 0)
    throw new RangeError("game event string is truncated");
  return reader.readNullTerminatedString(Math.min(maxBytes, availableBytes));
}

function requireWithin(
  reader: BitReader,
  bitLength: number,
  bits: number,
): void {
  if (reader.bitOffset + bits > bitLength)
    throw new RangeError("game event payload exceeds its declared bit length");
}

function requireBitLength(bytes: Uint8Array, bitLength: number): void {
  if (
    !Number.isSafeInteger(bitLength) ||
    bitLength < 0 ||
    bitLength > bytes.byteLength * 8
  )
    throw new RangeError(`game event bit length ${bitLength} is invalid`);
}

function positiveLimit(
  value: number | undefined,
  fallback: number,
  name: string,
): number {
  const result = value ?? fallback;
  if (!Number.isSafeInteger(result) || result <= 0)
    throw new RangeError(`${name} must be a positive safe integer`);
  return result;
}

function fieldType(id: number): GameEventFieldType {
  const entry = Object.entries(gameEventFieldTypes).find(
    ([, value]) => value === id,
  );
  if (!entry) throw new RangeError(`unsupported game event field type ${id}`);
  return entry[0] as GameEventFieldType;
}

function isRequiredName(value: string): value is RequiredGameEventName {
  return (
    value === "weapon_fire" ||
    value === "player_hurt" ||
    value === "player_death"
  );
}

File: crates/demo-source1-native/src/game_events.rs
use crate::bit_reader::BitReader;
use serde::Serialize;
use std::collections::{BTreeMap, HashSet};

#[derive(Clone, Copy, Debug, Eq, PartialEq, Serialize)]
pub enum FieldType {
    String,
    Float,
    Long,
    Short,
    Byte,
    Boolean,
    Uint64,
}
#[derive(Clone, Debug, Eq, PartialEq, Serialize)]
pub struct FieldSchema {
    pub name: String,
    pub field_type: FieldType,
}
#[derive(Clone, Debug, Eq, PartialEq, Serialize)]
pub struct EventSchema {
    pub id: u32,
    pub name: String,
    pub fields: Vec<FieldSchema>,
}
#[derive(Clone, Debug, PartialEq, Serialize)]
pub enum EventValue {
    Boolean(bool),
    Number(f64),
    String(String),
}
#[derive(Clone, Debug, PartialEq, Serialize)]
pub struct DecodedGameEvent {
    pub id: u32,
    pub name: String,
    pub fields: Vec<(String, EventValue)>,
    pub schema: EventSchema,
}
#[derive(Clone, Debug, PartialEq, Serialize)]
pub enum Availability<T> {
    Observed {
        value: T,
        event_id: u32,
        field: String,
    },
    Unavailable {
        reason: String,
    },
}
#[derive(Clone, Debug, PartialEq, Serialize)]
pub struct RequiredEvent {
    pub name: String,
    pub event_id: u32,
    pub tick: i32,
    pub actor_user_id: Availability<f64>,
    pub victim_user_id: Availability<f64>,
    pub attacker_user_id: Availability<f64>,
    pub weapon: Availability<String>,
    pub damage: Availability<f64>,
    pub health: Availability<f64>,
}
#[must_use]
pub fn project_required_event(event: &DecodedGameEvent, tick: i32) -> Option<RequiredEvent> {
    if !matches!(
        event.name.as_str(),
        "weapon_fire" | "player_hurt" | "player_death"
    ) {
        return None;
    }
    let numeric = |field: &str| match event
        .fields
        .iter()
        .find(|(name, _)| name == field)
        .map(|(_, v)| v)
    {
        Some(EventValue::Number(value)) => Availability::Observed {
            value: *value,
            event_id: event.id,
            field: field.into(),
        },
        _ => Availability::Unavailable {
            reason: format!("schema does not expose numeric {field}"),
        },
    };
    let string = |field: &str| match event
        .fields
        .iter()
        .find(|(name, _)| name == field)
        .map(|(_, v)| v)
    {
        Some(EventValue::String(value)) => Availability::Observed {
            value: value.clone(),
            event_id: event.id,
            field: field.into(),
        },
        _ => Availability::Unavailable {
            reason: format!("schema does not expose string {field}"),
        },
    };
    let unavailable = |reason: &str| Availability::Unavailable {
        reason: reason.into(),
    };
    Some(RequiredEvent {
        name: event.name.clone(),
        event_id: event.id,
        tick,
        actor_user_id: numeric("userid"),
        victim_user_id: if event.name == "weapon_fire" {
            unavailable("event has no victim role")
        } else {
            numeric("userid")
        },
        attacker_user_id: if event.name == "weapon_fire" {
            unavailable("event has no attacker role")
        } else {
            numeric("attacker")
        },
        weapon: string("weapon"),
        damage: if event.name == "player_hurt" {
            numeric("dmg_health")
        } else {
            unavailable("event has no damage field")
        },
        health: if event.name == "player_hurt" {
            numeric("health")
        } else {
            unavailable("event has no health field")
        },
    })
}
#[derive(Clone, Copy, Debug)]
pub struct EventLimits {
    pub max_events: usize,
    pub max_fields_per_event: usize,
    pub max_string_bytes: usize,
    pub max_message_bits: usize,
    pub max_total_schema_bytes: usize,
}
impl Default for EventLimits {
    fn default() -> Self {
        Self {
            max_events: 512,
            max_fields_per_event: 128,
            max_string_bytes: 4096,
            max_message_bits: 8 * 1024 * 1024,
            max_total_schema_bytes: 4 * 1024 * 1024,
        }
    }
}

pub fn decode_event_list(
    bytes: &[u8],
    bit_length: usize,
    event_count: usize,
    l: EventLimits,
) -> Result<Vec<EventSchema>, String> {
    validate(bytes, bit_length)?;
    if bit_length > l.max_message_bits {
        return Err("game event list bit limit exceeded".into());
    }
    if event_count > l.max_events {
        return Err(format!("game event count {event_count} is invalid"));
    }
    let mut r = BitReader::span(bytes, 0, bit_length).map_err(text)?;
    let mut ids = HashSet::new();
    let mut output = Vec::with_capacity(event_count);
    let mut schema_bytes = 0_usize;
    for _ in 0..event_count {
        let id = r.read_bits(9).map_err(text)?;
        if !ids.insert(id) {
            return Err(format!("duplicate game event id {id}"));
        }
        let name = r.read_latin1_z(l.max_string_bytes).map_err(text)?;
        if name.is_empty() {
            return Err(format!("game event {id} has an empty name"));
        }
        schema_bytes = schema_bytes.saturating_add(name.len());
        let mut fields = Vec::new();
        loop {
            let kind = r.read_bits(3).map_err(text)?;
            if kind == 0 {
                break;
            }
            if fields.len() >= l.max_fields_per_event {
                return Err(format!("game event {id} field limit exceeded"));
            }
            let field_type = field_type(kind)?;
            let field_name = r.read_latin1_z(l.max_string_bytes).map_err(text)?;
            schema_bytes = schema_bytes.saturating_add(field_name.len());
            if schema_bytes > l.max_total_schema_bytes {
                return Err("game event schema byte limit exceeded".into());
            }
            if field_name.is_empty() {
                return Err(format!("game event {id} has an empty field name"));
            }
            if fields.iter().any(|f: &FieldSchema| f.name == field_name) {
                return Err(format!("game event {id} repeats field {field_name}"));
            }
            fields.push(FieldSchema {
                name: field_name,
                field_type,
            });
        }
        output.push(EventSchema { id, name, fields });
    }
    if r.bit_offset() != bit_length {
        return Err(format!(
            "game event list consumed {} of {bit_length} bits",
            r.bit_offset()
        ));
    }
    Ok(output)
}
pub fn decode_event(
    bytes: &[u8],
    bit_length: usize,
    schemas: &BTreeMap<u32, EventSchema>,
    l: EventLimits,
) -> Result<DecodedGameEvent, String> {
    validate(bytes, bit_length)?;
    if bit_length > l.max_message_bits {
        return Err("game event bit limit exceeded".into());
    }
    let mut r = BitReader::span(bytes, 0, bit_length).map_err(text)?;
    let id = r.read_bits(9).map_err(text)?;
    let schema = schemas
        .get(&id)
        .ok_or_else(|| format!("game event {id} has no registered schema"))?
        .clone();
    let mut fields = Vec::with_capacity(schema.fields.len());
    for field in &schema.fields {
        fields.push((
            field.name.clone(),
            read_value(&mut r, field.field_type, l.max_string_bytes)?,
        ));
    }
    if r.bit_offset() != bit_length {
        return Err(format!(
            "game event {id} consumed {} of {bit_length} bits",
            r.bit_offset()
        ));
    }
    Ok(DecodedGameEvent {
        id,
        name: schema.name.clone(),
        fields,
        schema,
    })
}
#[derive(Debug, Default)]
pub struct GameEventDecoder {
    schemas: BTreeMap<u32, EventSchema>,
    limits: EventLimits,
}
impl GameEventDecoder {
    pub fn register(&mut self, bytes: &[u8], bits: usize, count: usize) -> Result<(), String> {
        self.schemas = decode_event_list(bytes, bits, count, self.limits)?
            .into_iter()
            .map(|s| (s.id, s))
            .collect();
        Ok(())
    }
    pub fn decode(&self, bytes: &[u8], bits: usize) -> Result<DecodedGameEvent, String> {
        decode_event(bytes, bits, &self.schemas, self.limits)
    }
    #[must_use]
    pub fn schemas(&self) -> &BTreeMap<u32, EventSchema> {
        &self.schemas
    }
}
fn read_value(r: &mut BitReader<'_>, kind: FieldType, max: usize) -> Result<EventValue, String> {
    Ok(match kind {
        FieldType::String => EventValue::String(r.read_latin1_z(max).map_err(text)?),
        FieldType::Float => EventValue::Number(f64::from(r.read_f32().map_err(text)?)),
        FieldType::Long => EventValue::Number(f64::from(
            i32::try_from(r.read_signed_bits(32).map_err(text)?).expect("32-bit signed value"),
        )),
        FieldType::Short => EventValue::Number(f64::from(
            i16::try_from(r.read_signed_bits(16).map_err(text)?).expect("16-bit signed value"),
        )),
        FieldType::Byte => EventValue::Number(f64::from(r.read_bits(8).map_err(text)?)),
        FieldType::Boolean => EventValue::Boolean(r.read_bool().map_err(text)?),
        FieldType::Uint64 => {
            let low = u64::from(r.read_bits(32).map_err(text)?);
            let high = u64::from(r.read_bits(32).map_err(text)?);
            EventValue::String(((high << 32) | low).to_string())
        }
    })
}
fn field_type(id: u32) -> Result<FieldType, String> {
    Ok(match id {
        1 => FieldType::String,
        2 => FieldType::Float,
        3 => FieldType::Long,
        4 => FieldType::Short,
        5 => FieldType::Byte,
        6 => FieldType::Boolean,
        7 => FieldType::Uint64,
        _ => return Err(format!("unsupported game event field type {id}")),
    })
}
fn validate(bytes: &[u8], bits: usize) -> Result<(), String> {
    if bits > bytes.len() * 8 {
        return Err(format!("game event bit length {bits} is invalid"));
    }
    Ok(())
}
#[allow(clippy::needless_pass_by_value)]
fn text(e: impl ToString) -> String {
    e.to_string()
}
#[cfg(test)]
mod tests {
    use super::*;
    #[test]
    fn rejects_unknown_schema() {
        let e = decode_event(&[0, 0], 9, &BTreeMap::new(), EventLimits::default()).unwrap_err();
        assert!(e.contains("no registered schema"));
    }
    #[test]
    fn empty_list_is_exact() {
        assert!(
            decode_event_list(&[], 0, 0, EventLimits::default())
                .unwrap()
                .is_empty()
        );
    }
}
